core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -638,6 +621,60 @@ trait OptimisticTransactionImpl extends TransactionalWrite
     }
   }
 
+  /**
+   * Ensure all native table features required by `currentMetadata` is supported by
+   * `currentProtocol`. This means these features must be either auto-update capable
+   * ([[FeatureAutomaticallyEnabledByMetadata.automaticallyUpdateProtocolOfExistingTables]] is set
+   * to `true`), or being supported in the same transaction via a table property.
+   */
+  private def assertTableFeaturesAutomaticallySupported(
+      currentProtocol: Protocol,
+      currentMetadata: Metadata): Unit = {
+    val (readerVersionFromTableConfOpt, writerVersionFromTableConfOpt) =
+      Protocol.getProtocolVersionsFromTableConf(currentMetadata.configuration)
+    val txnUpdatedProtocol = currentProtocol.merge(
+      Protocol(
+        readerVersionFromTableConfOpt.getOrElse(currentProtocol.minReaderVersion),
+        writerVersionFromTableConfOpt.getOrElse(currentProtocol.minWriterVersion)))
+    // Collect features that (1) are already supported by the current protocol or by a new
+    // protocol that the table will be updated to; (2) will be added to the protocol by this txn
+    // via `delta.feature.xxx` configs.
+    val supportedFeaturesInTxn =
+      TableFeatureProtocolUtils.getSupportedFeaturesFromConfigs(currentMetadata.configuration) ++
+        txnUpdatedProtocol.implicitlySupportedFeatures ++
+        txnUpdatedProtocol.readerAndWriterFeatureNames.flatMap(TableFeature.featureNameToFeature)
+    // Get the min set of features required by the metadata.
+    val (_, _, metadataEnabledFeaturesInTxn) =
+      Protocol.minProtocolComponentsFromAutomaticallyEnabledFeatures(spark, currentMetadata)
+    // Collect native features that are required by the metadata but can't be automatically
+    // activated.
+    val unsupportedNativeFeatures =
+      metadataEnabledFeaturesInTxn.diff(supportedFeaturesInTxn).filter {
+        case f: FeatureAutomaticallyEnabledByMetadata =>
+          !f.automaticallyUpdateProtocolOfExistingTables
+        case f =>
+          throw new RuntimeException(
+            f"Feature ${f.name} should not be here because " +
+            f"it is not activatable via metadata.")
```

Comment:
I changed it to `IllegalStateException`.

This code path is unexpected because all items in `metadataEnabledFeaturesInTxn` must be activatable via metadata, unless our code is wrong. Logging is not required as users are not gonna see this.